cerealize/cerealize.h

```diff
@@ -140,84 +140,124 @@ WithBaseType(const ENTRY& object) {
+class CerealFileAppenderBase {
+ public:
+  explicit CerealFileAppenderBase(const std::string& filename, bool append = true)
+       : fo_(filename, (append ? std::ofstream::app : std::ofstream::trunc) | std::ofstream::binary) {
+    long p = fo_.tellp();
+    if (p >= 0) {
+      initial_stream_position_ = static_cast<size_t>(p);
+    } else {
+      BRICKS_THROW(bricks::CerealizeFileStreamErrorException());
+    }
   }
-};
 
-template <>
-struct CerealStreamType<CerealFormat::JSON> {
-  typedef cereal::JSONInputArchive Input;
-  typedef cereal::JSONOutputArchive Output;
-  template <typename OSTREAM>
-  inline static Output CreateOutputArchive(OSTREAM& os) {
-    // TODO(dkorolev): Add NoIndent as a policy once that cereal change is merged.
-    return Output(os, cereal::JSONOutputArchive::Options::NoIndent());
+  inline size_t EntriesWritten() { return entries_written_; }
+  inline size_t BytesWritten() {
+    long p = fo_.tellp();
+    if (p >= 0) {
+      return static_cast<size_t>(p - initial_stream_position_);
+    } else {
+      BRICKS_THROW(bricks::CerealizeFileStreamErrorException());
+    }
   }
+
+ protected:
+  std::ofstream fo_;
+  size_t entries_written_ = 0;
+
+ private:
+  size_t initial_stream_position_;
+
+  CerealFileAppenderBase() = delete;
+  CerealFileAppenderBase(const CerealFileAppenderBase&) = delete;
+  void operator=(const CerealFileAppenderBase&) = delete;
+  CerealFileAppenderBase(CerealFileAppenderBase&&) = delete;
+  void operator=(CerealFileAppenderBase&&) = delete;
 };
 
-// `CerealFileAppender` appends cereal-ized records to a file.
-// The format is selected by a template parameter, defaults to binary. All formats are supported.
+// `CerealBinaryFileAppender` appends cereal-ized records to a file in binary format.
 // Writes are performed using templated `operator <<(const T& entry)`.
-// Is type `T` defines a typedef of `CEREAL_BASE_TYPE`, polymorphic serialization is used.
-template <CerealFormat T_CEREAL_FORMAT>
-class GenericCerealFileAppender {
+// If type `T` defines a typedef of `CEREAL_BASE_TYPE`, polymorphic serialization is used.
+class CerealBinaryFileAppender : public CerealFileAppenderBase {
  public:
-  explicit GenericCerealFileAppender(const std::string& filename, bool append = true)
-      : fo_(filename, (append ? std::ofstream::app : std::ofstream::trunc) | std::ofstream::binary),
-        so_(CerealStreamType<T_CEREAL_FORMAT>::CreateOutputArchive(fo_)) {}
+  explicit CerealBinaryFileAppender(const std::string& filename, bool append = true)
+      : CerealFileAppenderBase(filename, append),
+        so_(cereal::BinaryOutputArchive(fo_)) {}
 
   template <typename T>
-  typename std::enable_if<sizeof(typename T::CEREAL_BASE_TYPE) != 0, GenericCerealFileAppender&>::type
+  typename std::enable_if<sizeof(typename T::CEREAL_BASE_TYPE) != 0, CerealBinaryFileAppender&>::type
   operator<<(const T& entry) {
     so_(WithBaseType<typename T::CEREAL_BASE_TYPE>(entry));
+    entries_written_++;
```

Comment:
Nit: Google C++ Style Guide prefers pre-increment to post-increment. Long story, just grab a cactus :)